WordPress/gutenberg
# Query block: client-side pagination

## What?


Built on top of [#53733](https://github.com/WordPress/gutenberg/pull/53733). Supersedes [#44034](https://github.com/WordPress/gutenberg/pull/44034).

This PR adds an option to the Query block to enhance the pagination with client-side navigation.

## Why?


To improve the user experience.

## How?


The pagination links are intercepted using a directive (`data-wp-on--click`) which prevents the regular server-side navigation and uses the Interactivity API's `navigate` function instead. It also prefetches the next/previous pages on mouseenter.

## Tasks

There are still many tasks that need to be done before this can be considered ready. I'll probably add more as we progress and find out what needs to be done.

- [x] Only add `supports.interactivity` and enqueue the `view.js` file if the "Enhanced Pagination" option is enabled.
- [x] Add accessibility.
- [x] Scroll to the top of the Query block after navigation.
- [x] Add loading status when loading the next page takes more than 300ms.
- [x] Don't enqueue styles when enhanced pagination is not activated.

### Tasks left for subsequent PRs

- Display a popup when a user tries to enable the enhanced pagination and the Pagination block is not included inside the Query block. The user should be able to choose between canceling the action, or adding the Pagination block (added at the end) using buttons in the popup.

## Testing Instructions





You can use https://playground.wordpress.net/gutenberg.html to test this PR.

- Add or use a template with the Query block.
- Under the "User Experience" sidebar panel enable the "Enhanced Pagination" setting
- Go to the page/post.
- Use the pagination links.
- The page should not refresh, and the content of the new page should appear on the screen.

## Screenshots or screencast 


https://github.com/WordPress/gutenberg/assets/3305402/e64e7113-62eb-41e2-a67f-e38a0a163534
